domain/consensus/processes/blockvalidator/pruning_violation_proof_of_work_and_difficulty_test.go

```diff
@@ -0,0 +1,243 @@
+func solveBlockWithWrongPOW(block *externalapi.DomainBlock) *externalapi.DomainBlock {
+	targetDifficulty := difficulty.CompactToBig(block.Header.Bits())
+	headerForMining := block.Header.ToMutable()
+	initialNonce := uint64(0)
+	for i := initialNonce; i < math.MaxUint64; i++ {
+		headerForMining.SetNonce(i)
+		if !pow.CheckProofOfWorkWithTarget(headerForMining, targetDifficulty) {
+			block.Header = headerForMining.ToImmutable()
+			return block
+		}
+	}
+
+	panic("Failed to solve block! cannot find a invalid POW for the test")
+}
+
+func TestCheckParentHeadersExist(t *testing.T) {
+	testutils.ForAllNets(t, true, func(t *testing.T, params *dagconfig.Params) {
+		factory := consensus.NewFactory()
+
+		tc, teardown, err := factory.NewTestConsensus(params, false, "TestCheckBlockIsNotPruned")
+		if err != nil {
+			t.Fatalf("Error setting up consensus: %+v", err)
+		}
+		defer teardown(false)
+
+		orphanBlock, _, err := tc.BuildBlockWithParents([]*externalapi.DomainHash{params.GenesisHash}, nil, nil)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		parentHash := externalapi.NewDomainHashFromByteArray(&[externalapi.DomainHashSize]byte{}) // Non existing parent hash
+		orphanBlock.Header = blockheader.NewImmutableBlockHeader(
+			orphanBlock.Header.Version(),
+			[]*externalapi.DomainHash{
+				parentHash,
+			},
+			orphanBlock.Header.HashMerkleRoot(),
+			orphanBlock.Header.AcceptedIDMerkleRoot(),
+			orphanBlock.Header.UTXOCommitment(),
+			orphanBlock.Header.TimeInMilliseconds(),
+			orphanBlock.Header.Bits(),
+			orphanBlock.Header.Nonce(),
+		)
+
+		_, err = tc.ValidateAndInsertBlock(orphanBlock)
+		errMissingParents := &ruleerrors.ErrMissingParents{}
+		if !errors.As(err, errMissingParents) {
+			t.Fatalf("Unexpected error: %+v", err)
+		}
+
+		if !externalapi.HashesEqual(errMissingParents.MissingParentHashes, []*externalapi.DomainHash{parentHash}) {
+			t.Fatalf("unexpected missing parents %s", errMissingParents.MissingParentHashes)
+		}
+
+		invalidBlock, _, err := tc.BuildBlockWithParents([]*externalapi.DomainHash{params.GenesisHash}, nil, nil)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		invalidBlock.Transactions[0].Version = constants.MaxTransactionVersion + 1 // This should invalidate the block
+		invalidBlock.Header = blockheader.NewImmutableBlockHeader(
+			invalidBlock.Header.Version(),
+			invalidBlock.Header.ParentHashes(),
+			merkle.CalculateHashMerkleRoot(invalidBlock.Transactions),
+			orphanBlock.Header.AcceptedIDMerkleRoot(),
+			orphanBlock.Header.UTXOCommitment(),
+			orphanBlock.Header.TimeInMilliseconds(),
+			orphanBlock.Header.Bits(),
+			orphanBlock.Header.Nonce(),
+		)
+
+		_, err = tc.ValidateAndInsertBlock(invalidBlock)
+		if !errors.Is(err, ruleerrors.ErrTransactionVersionIsUnknown) {
+			t.Fatalf("Unexpected error: %+v", err)
+		}
+
+		invalidBlockHash := consensushashing.BlockHash(invalidBlock)
+
+		invalidBlockChild, _, err := tc.BuildBlockWithParents([]*externalapi.DomainHash{params.GenesisHash}, nil, nil)
```

Comment:
Why not build the block with `invalidBlockHash` directly?

Reply:
Because BuildBlockWithParents and BuildUTXOInvalidBlock requires valid parents so they can calc ghostdag etc. And I want a block with valid fields except this one.